sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java

```diff
@@ -486,14 +535,36 @@ private PDone createWrite(PCollection<T> input) {
     // initial ParDo.
     PCollection<FileResult> results;
     final PCollectionView<Integer> numShardsView;
+    @SuppressWarnings("unchecked")
     Coder<BoundedWindow> shardedWindowCoder =
         (Coder<BoundedWindow>) input.getWindowingStrategy().getWindowFn().windowCoder();
     if (computeNumShards == null && numShardsProvider == null) {
       numShardsView = null;
-      results =
-          input.apply(
-              "WriteBundles",
-              ParDo.of(windowedWrites ? new WriteWindowedBundles() : new WriteUnwindowedBundles()));
+      if (windowedWrites) {
+        TupleTag<FileResult> writtenRecordsTag =
+            new TupleTag<FileResult>("writtenRecordsTag") {};
```

Comment:
These anonymous TupleTag classes capture the enclosing transform so they'll blow up the serialized graph unnecessarily. I think the code will work even if you remove `{}` - if it won't, you'll probably just need to specify some coders, because all the `{}` does is capture the type information.